### PR TITLE
perf(mlx): dequantize the KV cache to the fused attention kernel at prefill

### PR DESCRIPTION
**Pull request by Lyxot**
## What this fixes

**Quantizing the KV cache on MLX today costs context length instead of buying it.**

An unquantized cache goes to `mx.fast.scaled_dot_product_attention`, which is fused and materializes no scores. There is no fused equivalent for a quantized cache, so `mlx_lm` and `mlx_vlm` fall back to a Python `quantized_scaled_dot_product_attention` that dequantizes the cache into a materialized `[B, HQ, L, S]` score matrix. At a fixed prefill chunk that matrix grows with cache depth -- about 130 MB per layer per 1024 tokens, against the 2.9 MB per 1024 tokens that quantizing saves -- and it dwarfs the cache the quantization shrank.

The effect a caller actually sees, on `mlx-community/Qwen3-4B-4bit` (36 layers) through `mlx_lm` at its default 2048-token prefill chunk. Peak is above the 2160 MB of weights, so it is the cache plus everything attention allocates to process the prompt:

| prompt | KV cache in bf16 | 4-bit KV cache, today | 4-bit KV cache, this PR |
|---|---|---|---|
| 8192 | 2267 MB | 3337 MB — **1.47x worse** | **1582 MB** |
| 16384 | 3388 MB | 5734 MB — **1.69x worse** | **1902 MB** |
| 32768 | 5670 MB | 9967 MB — **1.76x worse** | **2613 MB** |

So quantizing the cache to fit a longer context gives a *shorter* one: it raises peak memory by half to three quarters, and the machine runs out during prompt processing sooner than it would have with no quantization at all. With this PR quantizing does what it was meant to do -- 1.4x to 2.2x below bf16, and 2.1x to 3.8x below what quantizing costs today.

## How it works

The fix is the one `mlx_vlm` already uses for its TurboQuant caches: dequantize the cache and hand the fused kernel the copy.

`quantized_sdpa_over` wraps each runtime's own `quantized_scaled_dot_product_attention`. The wrapper compares the two transients per cached token -- `HQ * L` score elements against `HKV` rows of keys and values, counted at both widths because the prefix view a `QuantizedKVCache` hands over is compacted before it is dequantized -- and once the scores would cost more, it dequantizes and calls `mx.fast.scaled_dot_product_attention` over the copy. The depth cancels, so the query count alone decides: 82 queries for 32 heads over 8 at 128 wide and 4 bits, moving with the packing (98 at 8 bits, 76 at 2 bits / group 32). Below that the runtime's own function answers, unchanged, so every decode step returns its own bits.

The transient is then one bf16 copy of the cache, 4 MB per 1024 tokens of depth per layer, and only one layer's copy is live at a time. That is the whole-model effect in the table above.

`install_quantized_attention()` wraps the function in `mlx_lm.models.base` and `mlx_vlm.models.base`, whichever the load has already imported, and is idempotent. `FastMLXModel.from_pretrained` returns through a single helper that calls it after the load's own imports, so no load path can return a model still bound to the unwrapped function. If MLX ever ships `mx.fast.quantized_scaled_dot_product_attention`, the installer patches nothing and the runtimes keep their own.

## Memory and performance

Apple M3 Max, mlx 0.32.2, mlx-lm 0.31.3. Llama-3-8B attention shape: 32 query heads over 8 key/value heads, 128 wide, 4-bit cache at group size 64, `mask="causal"`. Peak is the transient above the settled baseline; both paths are timed in alternating rounds so drift falls on both equally, medians of nine.

Per layer, per call:

| queries | cache depth | peak, today | peak, this PR | | time, today | time, this PR | |
|---|---|---|---|---|---|---|---|
| 2048 | 8192 | 1072.0 MB | **64.0 MB** | 16.8x less | 39.2 ms | **22.8 ms** | 1.72x faster |
| 2048 | 32768 | 4192.1 MB | **160.0 MB** | 26.2x less | 206.5 ms | **129.9 ms** | 1.59x faster |
| 512 | 32768 | 1048.1 MB | **136.0 MB** | 7.7x less | 38.5 ms | **26.2 ms** | 1.47x faster |
| 512 | 131072 | 4168.5 MB | **520.0 MB** | 8.0x less | 224.0 ms | **114.5 ms** | 1.96x faster |

On a real cache prefix view add the packed cache once more (36 MB at 32K), for the compaction the dequantization does first. Decode is unchanged: below the threshold it is the runtime's own function.

Whole model, `mlx-community/Qwen3-4B-4bit`, 32 generated tokens, 2048-token prefill chunks, 4-bit cache at group size 64. Single runs on a laptop, so treat the times as indicative; the peaks are deterministic:

| prompt | prefill, bf16 cache | prefill, today | prefill, this PR | peak, today | peak, this PR |
|---|---|---|---|---|---|
| 8192 | 7.8 s | 9.7 s | **7.6 s** | 3337 MB | **1582 MB** |
| 16384 | 20.7 s | 26.2 s | **18.9 s** | 5734 MB | **1902 MB** |
| 32768 | 58.0 s | 111.2 s | **58.5 s** | 9967 MB | **2613 MB** |

Prefill runs at the speed of the bf16 cache and faster than the replaced path at every depth -- at 32K the replaced path's 10 GB peak puts the machine under memory pressure and it takes twice as long. `mx.fast.scaled_dot_product_attention` is a fused kernel; the replaced path is a dequantizing matmul, a masked softmax and a second matmul over a score matrix that does not fit in any cache.

## Accuracy

Agreement with the runtime's own function above the threshold: worst relative divergence 1.5e-2 over 18 cases (both runtimes, three mask forms, three bit widths), which is bf16 rounding in a different order. Below the threshold the result is the runtime's own, exactly.

## Deliberate divergences

The wrapper is a drop-in replacement -- same argument names, shapes and result dtypes, including the float32 the replaced path widens to when the queries and either cache disagree -- with two differences above the threshold, both inherited from the fused kernel and both the behaviour the same runtime already has for an unquantized cache:

1. **A query row whose mask admits no key gets the fused kernel's answer**, not the replaced path's average over every cached value (boolean mask) or NaN (additive). The runtimes build such rows only as batch padding, which they discard.
2. **The caller's queries survive the call.** The replaced path does `queries *= scale` in place; the fused kernel takes `scale` as an argument.

The mask goes to the fused kernel as given, which accepts every form the runtimes build (`"causal"`, boolean and additive arrays at `[L, S]`, `[B, 1, L, S]` and `[B, HQ, L, S]`). `mlx_vlm`'s own quantized function aligns `[B, 1, L, S]` masks against its 5-D grouped scores; the fused kernel needs no alignment, so a batched mask is handed over unchanged.

## History

An earlier revision of this PR shipped an `mx.fast.metal_kernel` flash-attention kernel that read the quantized cache in place, holding the transient flat at 113.5 MB whatever the depth. Measured whole-model it was both slower and larger than dequantizing to the fused kernel (1727 / 1954 / 2669 MB and 10.5 / 52.2 / 100.5 s at the three prompts above), because only one layer's transient is live at a time, so the flatness bought nothing a caller could see while the kernel's prefill ran 1.1x-1.3x slower than stock. Replaced by the ~90-line module here.

## Validation

```bash
python -m pytest tests/test_mlx_attention_metal.py -q
# 31 passed
```

Covers the routing tie at four head geometries (grouped, ungrouped, MLA-shaped, mixed dtype) and three packings; agreement with both runtimes' own functions on either side of the threshold under `"causal"`, boolean and additive masks, batched for `mlx_vlm`; the transient on a real `QuantizedKVCache` prefix view; result dtype against the runtime's for matched and mismatched query, key and value dtypes; the installer on both runtimes, twice; and that every return path of `from_pretrained` installs. A mutation sweep of 19 mutations of the module and the loader kills all 19.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

**Comment by danielhanchen** [comment]
This installs `quantized_sdpa` into `mlx_lm.models.base` and `mlx_vlm.models.base` for the whole process from every `FastMLXModel.from_pretrained` return, so the up to 1.30x prefill slowdown and the mask rank 3/4 axis change apply to callers who never turned on a quantized KV cache. Could you gate the rebind on the cache actually being quantized, or confirm why an unconditional install is safe, before I review the kernel itself?

**Comment by Lyxot** [comment]
@codex review

File: tests/test_mlx_attention_metal.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
+"""Quantized-KV attention over a dequantized copy, against the runtime's own path, on real Metal."""
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply tests/test_mlx_attention_metal.py]: Confirmed: the MLX job runs each test file as an explicit step and this file was not listed, so its tests never ran in CI. Added it as its own step in the Apple Silicon job, matching the neighbouring steps (`python -m pytest tests/test_mlx_attention_metal.py -v -rs`), folded into ca69dc3f.

**Comment by Lyxot** [comment]
@codex review

File: tests/test_mlx_attention_metal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_mlx_attention_metal.py]: Correct at the time, and already fixed: ca69dc3f adds a `tests/test_mlx_attention_metal.py` step to `.github/workflows/mlx-ci.yml` (it landed about nine minutes after this comment). The Linux lanes deliberately do not take it, since the transient and routing assertions want real Metal.

Worth recording that the gap you point at was real for a different reason, though. That job is opt-in behind the `mlx` label, so on this PR it had never run at all. I replicated the branch onto a staging repo and labelled it, and the suite failed there: `mlx-vlm<0.7.0` under this repo's `transformers<=5.5.0` cap resolves to 0.6.4, which predates `align_attention_mask_to_scores`, so the `mlx_vlm-bool` cases raised a broadcast error inside the reference call. Fixed in 405d5796 along with two other items. I have added the `mlx` label here so the job runs on this PR from now on.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/attention.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        # An additive mask that does not promote to the result is rejected outright by the fused
+        # kernel, where the runtime's `scores += mask` widens the scores to it and carries on.
+        dtype = _result_dtype(queries, q_keys, q_values)
+        if mx.result_type(mask.dtype, dtype) != dtype:
+            return False
+    head_dim = q_keys[1].shape[-1] * group_size
+    value_dim = q_values[1].shape[-1] * group_size
+    L = queries.shape[-2]
+    if L > 8:
+        return head_dim == value_dim and head_dim in _FUSED_FULL_HEAD_DIMS
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/attention.py:64]: Checked this against mlx 0.32.1 and the premise does not hold: mlx promotes Q/K/V itself, before eligibility is decided, so a mixed-dtype call does not fall back.

`mlx/fast.cpp:704-715`:

```cpp
auto final_type = result_type(queries, keys, values);
...
auto q = astype(queries, final_type, s);
auto k = astype(keys, final_type, s);
auto v = astype(values, final_type, s);
```

`ScaledDotProductAttention::use_fallback` is then called with those promoted arrays at `fast.cpp:828-838`, and its body (`mlx/backend/metal/scaled_dot_product_attention.cpp:596-647`) tests only `is_training`, `output_logsumexp`, the device, the head dims, the sequence lengths, the GQA factor and the mask form. There is no dtype condition in it, which is why the eligibility check here is dimensions-only.

Confirmed by running it: bf16 queries against a float32 key/value cache returns float32 with no error, rather than raising or degrading. The residual cost is one float32 copy of the queries, which is `HQ * L * D` and does not grow with cache depth, so it cannot overtake the `[B, HQ, L, S]` scores the route avoids. The byte model already uses the widened `dtype.size` on both sides of the comparison for this case.

The dtype constraint that is real is the one on the mask, and that is already gated: `fast.cpp:796` rejects a mask that does not promote to `final_type`, and `fused_kernel_exists` declines the route for exactly that case.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/attention.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    patched = []
+    for module_path in _PATCH_TARGETS:
+        module = sys.modules.get(module_path)
+        if module is None or getattr(module, _PATCH_FLAG, False):
+            continue
+        unfused = getattr(module, "quantized_scaled_dot_product_attention", None)
+        if unfused is None:
+            continue
+        module.quantized_scaled_dot_product_attention = quantized_sdpa_over(unfused)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/attention.py:115]: No architecture module binds that symbol, so there is no reference for the patch to miss. Grepping the whole of both pinned trees gives four hits, all in the two base modules:

```
mlx_lm/models/base.py:64:  def quantized_scaled_dot_product_attention(
mlx_lm/models/base.py:120:     return quantized_scaled_dot_product_attention(
mlx_vlm/models/base.py:228: def quantized_scaled_dot_product_attention(
mlx_vlm/models/base.py:323:    return quantized_scaled_dot_product_attention(
```

Definition and single call site. There is no `from ... import quantized_scaled_dot_product_attention` anywhere in mlx-lm 0.31.3 or mlx-vlm 0.6.4. What the architectures import is the dispatcher, `from .base import scaled_dot_product_attention`, and that function resolves the quantized one as a module global at call time, so rebinding the module attribute reaches every caller however early the architecture was imported.

Ran it to be sure, importing the model modules first and installing afterwards, the order `_finish_load` produces:

```
llama binds     : ['scaled_dot_product_attention']
qwen3 binds     : ['scaled_dot_product_attention']
patched         : ('mlx_lm.models.base',)
route taken     : ['FUSED'] -> a pre-patch model reference DID reach the wrapper
```

The call went through `llama.scaled_dot_product_attention` captured before the patch, against a real `QuantizedKVCache`, and still took the fused path. Installing earlier would also not work: the installer only patches runtimes already in `sys.modules`, so running it before the architectures are imported is the case that would miss them.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/attention.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    packed = sum(part.shape[-1] * part.dtype.size for part in q_cache)
+    return q_cache[1].shape[-1] * group_size * dtype.size + packed
+
+
+def dequantizing_is_smaller(queries, q_keys, q_values, group_size):
+    """Scores cost `HQ * L` per cached token, the copy `HKV` rows; the depth cancels."""
+    dtype = _result_dtype(queries, q_keys, q_values)
+    HQ, L = queries.shape[-3], queries.shape[-2]
+    HKV = q_keys[0].shape[-3]
+    scores = HQ * L * dtype.size
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/attention.py:42]: The arithmetic is right and I reproduced your numbers exactly. Reproducing the runtime's grouped reshape with fp16 queries, fp16 key scales and bf16 value scales:

```
actual SCORE dtype  = mlx.core.float16
actual RESULT dtype = mlx.core.float32  | _result_dtype() = mlx.core.float32
routes fused from L=74; with true float16 scores it should be L=147
```

So the predicate does price fp16 scores as float32 across that window, which is your 75 to 148.

Rejecting on reachability rather than on the analysis. The mismatch needs `k_scales.dtype == queries.dtype != v_scales.dtype`, and no cache produces it: `QuantizedKVCache.update_and_fetch` quantizes keys and values from the two projections of one hidden state, so their scales always land in the same dtype.

```
real QuantizedKVCache -> k_scales bfloat16 / v_scales bfloat16
```

Every other combination is already priced correctly, because the scores promote along with the result: queries wider than the cache give float32 scores and a float32 result, and a matched cache gives bf16 for both. Only a key/value split, which cannot occur, separates them. The parametrisation you are pointing at is a dtype-contract test on `dequantized_sdpa`, asserting the result dtype matches the runtime's; it is not a cache shape a load can build.

Worth contrasting with the head-dim guard added in 405d5796, which I did take: that one is reachable today (Qwen3-Next is head_dim 256, and `kv_bits` reaches `stream_generate` through `FastMLXModel.generate`) and it costs the score matrix plus the copy. This one costs nothing on any input a model can produce.

**Comment by danielhanchen** [comment]
@codex review
